// relaen/dist/core/transaction/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const relaenmanager_1 = require("../relaenmanager");
const logger_1 = require("../logger");
const mysqltransaction_1 = require("./mysqltransaction");
/**
 * 事务基类
 */
class Transaction {
    constructor(connection) {
        this.connection = connection;
    }
    /**
     * 事务开始
     */
    async begin() {
        logger_1.Logger.console('[Relaen]:Transaction is started!');
        switch (relaenmanager_1.RelaenManager.dialect) {
            case 'mysql':
                this.tr = new mysqltransaction_1.MysqlTransaction();
                break;
            case 'oracle':
            case 'mssql':
        }
        await this.tr.begin(this.connection);
    }
    /**
     * 提交
     */
    async commit() {
        if (this.tr && this.connection.connected) {
            await this.tr.commit(this.connection);
        }
        logger_1.Logger.console('[Relaen]:Transaction is commited!');
    }
    /**
     * 回滚
     */
    async rollback() {
        if (this.tr && this.connection.connected) {
            await this.tr.rollback(this.connection);
        }
        logger_1.Logger.console('[Relaen]:Transaction is rollbacked!');
    }
}
exports.Transaction = Transaction;
