// relaen/dist/core/relaenmanager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelaenManager = void 0;
const connectionmanager_1 = require("./connectionmanager");
const entityfactory_1 = require("./entityfactory");
const errorfactory_1 = require("./errorfactory");
/**
 * relaen 框架管理器
 */
class RelaenManager {
    /**
     * 初始化
     * @param cfg   配置文件名或配置对象
     */
    static init(cfg) {
        if (typeof cfg === 'string') {
            cfg = this.parseFile(cfg);
        }
        if (typeof cfg !== 'object') {
            throw errorfactory_1.ErrorFactory.getError('0001');
        }
        this.dialect = cfg.dialect || 'mysql';
        this.debug = cfg.debug || false;
        this.cache = cfg.cache === false ? false : true;
        connectionmanager_1.ConnectionManager.init(cfg);
        //加载实体
        for (let path of cfg.entities) {
            entityfactory_1.EntityFactory.addEntities(path);
        }
    }
    /**
     * @exclude
     * 解析实例配置文件
     * @param path      文件路径
     */
    static parseFile(path) {
        path = path || 'relaen.json';
        //读取文件
        let jsonStr = require('fs').readFileSync(path, 'utf-8');
        let json;
        try {
            json = JSON.parse(jsonStr);
        }
        catch (e) {
            throw "error config file!";
        }
        return json;
    }
}
exports.RelaenManager = RelaenManager;
