// relaen/dist/core/sqlexecutor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqlExecutor = void 0;
const relaenmanager_1 = require("./relaenmanager");
const logger_1 = require("./logger");
const errorfactory_1 = require("./errorfactory");
/**
 * sql执行器
 */
class SqlExecutor {
    /**
     * 执行mysql sql语句
     * @param connection    db connection
     * @param sql           待执行sql
     * @param params        参数数组
     * @param start         开始记录行
     * @param limit         最大记录行
     * @returns             执行结果或undefined
     */
    static async exec(em, sql, params, start, limit) {
        sql = sql.trim();
        //sql类型：0:查询 1:增删改
        let sqlType = ['insert', 'update', 'delete'].includes(sql.substr(0, 6).toLowerCase()) ? 1 : 0;
        //缓存key，构建方式：sql_paramsvaluestring
        let key;
        //结果
        let result;
        if (sqlType === 0) { //查询可从缓存中获取
            sql = this.handleStartAndLimit(sql, start, limit);
            key = sql;
            //构造缓存key
            if (params) {
                key += '_' + JSON.stringify(params);
            }
            //从缓存获取
            result = em.getFromCache(key);
            //缓存中存在，则直接返回
            if (result) {
                return result;
            }
        }
        //打印sql
        logger_1.Logger.console("[Relaen execute sql]:\"" + sql + "\"");
        //打印参数
        if (params) {
            logger_1.Logger.console("Parameters is " + JSON.stringify(params));
        }
        try {
            switch (relaenmanager_1.RelaenManager.dialect) {
                case 'mysql':
                    result = await this.execMysql(em.connection, sql, params);
                    break;
                case 'oracle':
                    result = await this.execOracle(em.connection, sql, params);
                    break;
                case 'mssql':
                    result = await this.execMssql(em.connection, sql, params);
                    break;
            }
            //执行增删改，则清空cache
            if (sqlType === 1) {
                em.clearCache();
            }
            else { //添加到缓存
                em.addToCache(key, result);
            }
        }
        catch (e) {
            throw ("[Relaen execute sql] Error:\"" + e.message + "\"");
        }
        logger_1.Logger.console("[Relaen execute sql]:\"OK\"");
        return result;
    }
    /**
     * 执行mysql sql语句
     * @param connection    db connection
     * @param sql           待执行sql
     * @param params        参数数组
     * @param start         开始记录行
     * @param limit         最大记录行
     */
    static async execMysql(connection, sql, params) {
        if (sql.length < 6) {
            throw errorfactory_1.ErrorFactory.getError("0002", [sql]);
        }
        let r = await new Promise((resolve, reject) => {
            connection.conn.query(sql, params, (error, results, fields) => {
                if (error) {
                    reject(error);
                }
                resolve(results);
            });
        });
        if (r.insertId) {
            return r.insertId;
        }
        return r;
    }
    /**
     * 执行oracle sql语句
     * @param sql       待执行sql
     * @param params    参数数组
     */
    static async execOracle(connection, sql, params) {
    }
    /**
     * 执行mssql sql语句
     * @param sql       待执行sql
     * @param params    参数数组
     */
    static async execMssql(connection, sql, params) {
    }
    /**
     * 处理记录起始记录索引和记录数
     * @param sql       sql
     * @param start     开始索引
     * @param limit     记录数
     * @returns         处理后的sql
     * @since           0.2.0
     */
    static handleStartAndLimit(sql, start, limit) {
        if (!Number.isInteger(start) || start < 0 || !Number.isInteger(limit) || limit <= 0) {
            return sql;
        }
        switch (relaenmanager_1.RelaenManager.dialect) {
            case 'mysql':
                return sql + ' limit ' + start + ',' + limit;
            case 'oracle':
                return sql;
            case 'mssql':
                return sql;
        }
    }
}
exports.SqlExecutor = SqlExecutor;
