// noomi/dist/core/database/mssqltransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MssqlTransaction = void 0;
const noomitransaction_1 = require("./noomitransaction");
const connectionmanager_1 = require("./connectionmanager");
const dbmanager_1 = require("./dbmanager");
/**
 * mssql 事务类
 */
class MssqlTransaction extends noomitransaction_1.NoomiTransaction {
    /**
     * 构造器
     * @param id            事务id
     * @param connection    所属连接
     * @param type          事务类型
     */
    constructor(id, connection, type) {
        super(id, connection, type);
        let cm = dbmanager_1.DBManager.getConnectionManager();
        let pool = cm.pool;
        this.tr = new cm.dbMdl.Transaction(pool);
    }
    /**
     * 开始事务
     */
    async begin() {
        if (!this.connection) {
            this.connection = await connectionmanager_1.getConnection();
        }
        await this.tr.begin();
    }
    /**
     * 事务提交
     */
    async commit() {
        await this.tr.commit();
    }
    /**
     * 事务回滚
     */
    async rollback() {
        await this.tr.rollback();
    }
}
exports.MssqlTransaction = MssqlTransaction;
