// noomi/dist/core/database/mssqlconnectionmanager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MssqlConnectionManager = void 0;
const transactionmanager_1 = require("./transactionmanager");
/**
 * mssql连接管理器
 * @remarks
 * mssql直接启动连接池，不需要单独配置
 */
class MssqlConnectionManager {
    /**
     * 构造器
     * @param cfg 配置对象 {usePool:使用连接池,useTransaction:是否启用事务机制,其它配置参考options属性说明}
     */
    constructor(cfg) {
        this.dbMdl = require('mssql');
        delete cfg.useTransaction;
        this.options = cfg;
        this.pool = new this.dbMdl.ConnectionPool(this.options);
    }
    /**
     * 获取连接
     * @returns mssql request对象
     */
    async getConnection() {
        let conn = await transactionmanager_1.TransactionManager.getConnection();
        if (conn) {
            return conn;
        }
        let tr = transactionmanager_1.TransactionManager.get(false);
        let co;
        if (tr) {
            co = new this.dbMdl.Request(tr.tr);
        }
        else {
            let c = await this.pool.connect();
            co = c.request();
        }
        return co;
    }
    /**
     * 释放连接
     * @param request mssql request对象
     */
    async release(request) {
        if (!request) {
            return;
        }
        request.connection.close({ drop: false });
    }
}
exports.MssqlConnectionManager = MssqlConnectionManager;
