// noomi/dist/core/database/mysqlconnectionmanager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MysqlConnectionManager = void 0;
const transactionmanager_1 = require("./transactionmanager");
/**
 * mysql连接管理器
 */
class MysqlConnectionManager {
    /**
     * 构造器
     * @param cfg 配置对象 {usePool:使用连接池,useTransaction:是否启用事务机制,其它配置参考options属性说明}
     */
    constructor(cfg) {
        this.dbMdl = require('mysql');
        this.usePool = cfg.usePool || false;
        if (this.usePool) {
            delete cfg.usePool;
            delete cfg.useTransaction;
            this.pool = this.dbMdl.createPool(cfg);
        }
        this.options = cfg;
    }
    /**
     * 获取连接
     * @returns connection 对象
     */
    async getConnection() {
        let conn = await transactionmanager_1.TransactionManager.getConnection();
        if (conn) {
            return conn;
        }
        if (this.pool) {
            return new Promise((resolve, reject) => {
                this.pool.getConnection((err, conn) => {
                    if (err) {
                        reject(err);
                    }
                    resolve(conn);
                });
            });
        }
        else {
            return await this.dbMdl.createConnection(this.options);
        }
    }
    /**
     * 释放连接
     * @param conn connection对象
     */
    async release(conn) {
        if (!conn) {
            return;
        }
        if (this.pool) {
            conn.release(err => {
                console.log(err);
            });
        }
        else {
            conn.end(err => {
                console.log(err);
            });
        }
    }
}
exports.MysqlConnectionManager = MysqlConnectionManager;
