// noomi/dist/core/main/aopfactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AopPointcut = exports.AopFactory = void 0;
const instancefactory_1 = require("./instancefactory");
const aopproxy_1 = require("./aopproxy");
const errorfactory_1 = require("../tools/errorfactory");
const transactionmanager_1 = require("../database/transactionmanager");
const util_1 = require("../tools/util");
const application_1 = require("../tools/application");
/**
 * aop 切点类
 */
class AopPointcut {
    /**
     * 构造器
     * @param id            切点id(唯一)
     * @param expressions   该切点对应的表达式数组，表达式为正则表达式串
     */
    constructor(id, expressions) {
        /**
         * 表达式数组（正则表达式）
         */
        this.expressions = [];
        /**
         * 通知数组
         */
        this.advices = [];
        this.id = id;
        if (!expressions) {
            throw new errorfactory_1.NoomiError("2001");
        }
        if (!Array.isArray(expressions)) {
            expressions = [expressions];
        }
        expressions.forEach((item) => {
            if (typeof item !== 'string') {
                throw new errorfactory_1.NoomiError("2001");
            }
            this.expressions.push(util_1.Util.toReg(item));
        });
    }
    /**
     * 匹配方法是否满足表达式
     * @param instanceName  实例名
     * @param methodName    待检测方法
     * @returns             匹配返回true，否则返回false
     */
    match(instanceName, methodName) {
        for (let i = 0; i < this.expressions.length; i++) {
            if (this.expressions[i].test(instanceName + '.' + methodName)) {
                return true;
            }
        }
        return false;
    }
    /**
     * 给切点添加通知
     * @param advice    通知对象
     */
    addAdvice(advice) {
        this.advices.push(advice);
    }
}
exports.AopPointcut = AopPointcut;
/**
 * Aop工厂
 * 用于管理所有切面、切点
 */
class AopFactory {
    /**
     * 添加一个切面
     * @param cfg   切面对象
     */
    static addAspect(cfg) {
        if (this.aspects.has(cfg.instance)) {
            throw new errorfactory_1.NoomiError("2005", cfg.instance);
        }
        //连接点
        if (Array.isArray(cfg.advices)) {
            cfg.advices.forEach((item) => {
                if (!this.pointcuts.has(item.pointcut_id)) {
                    throw new errorfactory_1.NoomiError("2002", item.pointcut_id);
                }
                //设置实例或实例名
                item.instance = cfg.instance;
                //添加到pointcut的aop数组(是否需要重复检测，待考虑)
                this.addAdvice(item);
            });
        }
        this.aspects.set(cfg.instance, cfg);
    }
    /**
     * 添加切点
     * @param id            切点id
     * @param expressions   方法匹配表达式数组
     */
    static addPointcut(id, expressions) {
        this.pointcuts.set(id, new AopPointcut(id, expressions));
        //增加代理处理，此方法只在InstanceFactory init后执行
        instancefactory_1.InstanceFactory.addAfterInitOperation(this.updMethodProxy, this);
    }
    /**
     * 为切点添加表达式
     * @param pointcutId    切点id
     * @param expression    表达式或数组
     */
    static addExpression(pointcutId, expression) {
        if (!this.pointcuts.has(pointcutId)) {
            throw new errorfactory_1.NoomiError("2002", pointcutId);
        }
        let pc = this.pointcuts.get(pointcutId);
        if (!Array.isArray(expression)) {
            let reg = util_1.Util.toReg(expression);
            pc.expressions.push(reg);
        }
        else {
            expression.forEach(item => {
                let reg = util_1.Util.toReg(item);
                pc.expressions.push(reg);
            });
        }
        //把更新方法代理加入实例工厂后处理
        instancefactory_1.InstanceFactory.addAfterInitOperation(this.updMethodProxy, this);
    }
    /**
     * 为切点添加一个通知
     * @param advice 通知配置
     */
    static addAdvice(advice) {
        let pc = AopFactory.getPointcutById(advice.pointcut_id);
        if (!pc) {
            throw new errorfactory_1.NoomiError("2002", advice.pointcut_id);
        }
        pc.addAdvice(advice);
    }
    /**
     * @exclude
     * 解析文件
     * @param path  文件路径
     */
    static parseFile(path) {
        //读取文件
        let jsonStr = application_1.App.fs.readFileSync(path, 'utf-8');
        let json = null;
        try {
            json = application_1.App.JSON.parse(jsonStr);
        }
        catch (e) {
            throw new errorfactory_1.NoomiError("2000") + '\n' + e;
        }
        this.init(json);
    }
    /**
     * 初始化切面工厂
     * @param config 配置对象，包含切点集合、切面集合(含通知集合)
     */
    static init(config) {
        //切点数组
        if (Array.isArray(config.pointcuts)) {
            config.pointcuts.forEach((item) => {
                this.addPointcut(item.id, item.expressions);
            });
        }
        //切面数组
        if (Array.isArray(config.aspects)) {
            config.aspects.forEach((item) => {
                this.addAspect(item);
            });
        }
    }
    /**
     * 更新aop匹配的方法代理，为所有aop匹配的方法设置代理
     */
    static updMethodProxy() {
        if (!this.pointcuts || this.pointcuts.size === 0) {
            return;
        }
        //遍历pointcut
        let pc;
        for (pc of this.pointcuts.values()) {
            let reg;
            //遍历expression
            for (reg of pc.expressions) {
                this.addProxyByExpression(reg);
            }
        }
    }
    /**
     * 通过正则式给方法加代理
     * @param expr          表达式正则式
     */
    static addProxyByExpression(expr) {
        //遍历instance factory设置aop代理
        let insFac = instancefactory_1.InstanceFactory.getFactory();
        for (let insName of insFac.keys()) {
            //先检测instanceName
            let instance = instancefactory_1.InstanceFactory.getInstance(insName);
            if (instance) {
                Object.getOwnPropertyNames(instance.__proto__).forEach(key => {
                    //给方法设置代理，constructor 不需要代理
                    if (key === 'constructor' || typeof (instance[key]) !== 'function') {
                        return;
                    }
                    let method = insName + '.' + key;
                    //已代理过，不再代理
                    if (this.proxyMethodMap.has(method)) {
                        return;
                    }
                    //实例名+方法符合aop正则表达式
                    if (expr.test(method)) {
                        instance[key] = aopproxy_1.AopProxy.invoke(insName, key, instance[key], instance);
                        this.proxyMethodMap.set(method, true);
                    }
                });
            }
        }
    }
    /**
     * 获取切点
     * @param instanceName  实例名
     * @param methodName    方法名
     * @returns             切点数组
     */
    static getPointcut(instanceName, methodName) {
        // 遍历iterator
        let a = [];
        for (let p of this.pointcuts.values()) {
            if (p.match(instanceName, methodName)) {
                a.push(p);
            }
        }
        return a;
    }
    /**
     * 根据id获取切点
     * @param pointcutId    切点id
     * @returns             切点对象
     */
    static getPointcutById(pointcutId) {
        return this.pointcuts.get(pointcutId);
    }
    /**
     * 获取advices
     * @param instanceName  实例名
     * @param methodName    方法名
     * @return              {
     *                          before:[{instance:切面实例,method:切面方法},...]
     *                          after:[{instance:切面实例,method:切面方法},...]
     *                          return:[{instance:切面实例,method:切面方法},...]
     *                          throw:[{instance:切面实例,method:切面方法},...]
     *                      }
     */
    static getAdvices(instanceName, methodName) {
        let pointcuts = this.getPointcut(instanceName, methodName);
        if (pointcuts.length === 0) {
            return null;
        }
        let beforeArr = [];
        let afterArr = [];
        let throwArr = [];
        let returnArr = [];
        let pointcut;
        let hasTransaction = false;
        for (pointcut of pointcuts) {
            if (pointcut.id === transactionmanager_1.TransactionManager.pointcutId) {
                hasTransaction = true;
                continue;
            }
            pointcut.advices.forEach(aop => {
                let ins = typeof aop.instance === 'string' ?
                    instancefactory_1.InstanceFactory.getInstance(aop.instance) : aop.instance;
                switch (aop.type) {
                    case 'before':
                        beforeArr.push({
                            instance: ins,
                            method: aop.method
                        });
                        return;
                    case 'after':
                        afterArr.push({
                            instance: ins,
                            method: aop.method
                        });
                        return;
                    case 'around':
                        beforeArr.push({
                            instance: ins,
                            method: aop.method
                        });
                        afterArr.push({
                            instance: ins,
                            method: aop.method
                        });
                        return;
                    case 'after-return':
                        returnArr.push({
                            instance: ins,
                            method: aop.method
                        });
                        return;
                    case 'after-throw':
                        throwArr.push({
                            instance: ins,
                            method: aop.method
                        });
                }
            });
        }
        return {
            hasTransaction: hasTransaction,
            before: beforeArr,
            after: afterArr,
            throw: throwArr,
            return: returnArr
        };
    }
}
exports.AopFactory = AopFactory;
/**
 * 切面map，用于存储所有切面
 */
AopFactory.aspects = new Map();
/**
 * 切点map，用于存储所有切点
 */
AopFactory.pointcuts = new Map();
/**
 * 已代理方法map，键为instanctName.methodName，避免重复代理
 * @since 0.4.4
 */
AopFactory.proxyMethodMap = new Map();
